File: python/wallet_flow_shadow.py
"""Read-only aggregate Polymarket wallet-flow telemetry for offline research."""

from __future__ import annotations

import logging
import math
import time
from collections import defaultdict
from typing import Callable, Optional

import requests

from config import BotConfig
from models import MarketInfo

log = logging.getLogger("bot.wallet_flow_shadow")


class WalletFlowShadow:
    def __init__(self, config: BotConfig, now_fn: Callable[[], float] = time.time):
        self.config = config
        self._now_fn = now_fn

    def lookup(self, market: MarketInfo) -> Optional[dict]:
        """Fetch public trades and return anonymous market-level aggregates."""
        now = int(self._now_fn())
        window_minutes = max(1, self.config.wallet_flow_window_minutes)
        start = now - window_minutes * 60
        try:
            response = requests.get(
                f"{self.config.wallet_flow_api_host.rstrip('/')}/trades",
                params={
                    "market": market.condition_id,
                    "start": start,
                    "end": now,
                    "limit": max(1, min(self.config.wallet_flow_trades_limit, 10_000)),
                    "takerOnly": "true",
                },
                timeout=20,
            )
            response.raise_for_status()
            payload = response.json()
            trades = payload if isinstance(payload, list) else []
            result = self._aggregate(trades, market.condition_id, start, now, window_minutes)
            log.info(
                "Wallet-flow shadow: %d trades, $%.2f volume, imbalance=%+.2f",
                result["trade_count"], result["gross_volume_usd"], result["flow_imbalance"],
            )
            return result
        except Exception as exc:
            log.warning("Wallet-flow shadow lookup failed: %s", exc)
            return None

    def _aggregate(
        self, trades: list[dict], condition_id: str, start: int, end: int, window_minutes: int
    ) -> dict:
        yes_volume = 0.0
        no_volume = 0.0
        large_volume = 0.0
        trade_count = 0
        large_count = 0
        wallet_volume: dict[str, float] = defaultdict(float)
        threshold = max(0.0, self.config.wallet_flow_large_trade_usd)

        for trade in trades:
            if str(trade.get("conditionId", "")).lower() != condition_id.lower():
                continue
            timestamp = _number(trade.get("timestamp"))
            size = _number(trade.get("size"))
            price = _number(trade.get("price"))
            if timestamp < start or timestamp > end or size <= 0 or price <= 0:
                continue
            notional = size * price
            if not math.isfinite(notional) or notional <= 0:
                continue
            side = str(trade.get("side", "")).upper()
            outcome = str(trade.get("outcome", "")).upper()
            yes_direction = (outcome == "YES" and side == "BUY") or (outcome == "NO" and side == "SELL")
            no_direction = (outcome == "NO" and side == "BUY") or (outcome == "YES" and side == "SELL")
            if not yes_direction and not no_direction:
                continue

            trade_count += 1
            if yes_direction:
                yes_volume += notional
            else:
                no_volume += notional
            wallet = str(trade.get("proxyWallet", "")).strip()
            if wallet:
                wallet_volume[wallet] += notional
            if notional >= threshold:
                large_count += 1
                large_volume += notional

        gross = yes_volume + no_volume
        return {
            "window_minutes": window_minutes,
            "trade_count": trade_count,
            "wallet_count": len(wallet_volume),
            "gross_volume_usd": gross,
            "yes_direction_volume_usd": yes_volume,
            "no_direction_volume_usd": no_volume,
            "net_yes_flow_usd": yes_volume - no_volume,
            "flow_imbalance": (yes_volume - no_volume) / gross if gross else 0.0,
            "top_wallet_share": max(wallet_volume.values(), default=0.0) / gross if gross else 0.0,
            "large_trade_count": large_count,
            "large_trade_share": large_volume / gross if gross else 0.0,
            "observed_at": end,
        }


def _number(value) -> float:
    try:
        number = float(value or 0)
        return number if math.isfinite(number) else 0.0
    except (TypeError, ValueError):
        return 0.0

File: dotnet/PolymarketBot/Services/WalletFlowShadow.cs
using System.Text.Json;
using Microsoft.Extensions.Logging;
using PolymarketBot.Models;

namespace PolymarketBot.Services;

public sealed record WalletFlowReference(
    int WindowMinutes,
    int TradeCount,
    int WalletCount,
    double GrossVolumeUsd,
    double YesDirectionVolumeUsd,
    double NoDirectionVolumeUsd,
    double NetYesFlowUsd,
    double FlowImbalance,
    double TopWalletShare,
    int LargeTradeCount,
    double LargeTradeShare,
    long ObservedAt);

public sealed class WalletFlowShadow
{
    private readonly BotConfig _config;
    private readonly HttpClient _http;
    private readonly ILogger<WalletFlowShadow> _log;
    private readonly Func<DateTimeOffset> _now;

    public WalletFlowShadow(BotConfig config, HttpClient http, ILogger<WalletFlowShadow> log,
        Func<DateTimeOffset>? now = null)
    {
        _config = config;
        _http = http;
        _log = log;
        _now = now ?? (() => DateTimeOffset.UtcNow);
    }

    public async Task<WalletFlowReference?> LookupAsync(MarketInfo market, CancellationToken ct = default)
    {
        var now = _now().ToUnixTimeSeconds();
        var windowMinutes = Math.Max(1, _config.WalletFlowWindowMinutes);
        var start = now - windowMinutes * 60L;
        var limit = Math.Clamp(_config.WalletFlowTradesLimit, 1, 10_000);
        var url = $"{_config.WalletFlowApiHost.TrimEnd('/')}/trades" +
                  $"?market={Uri.EscapeDataString(market.ConditionId)}&start={start}&end={now}" +
                  $"&limit={limit}&takerOnly=true";
        try
        {
            using var response = await _http.GetAsync(url, ct);
            response.EnsureSuccessStatusCode();
            using var document = JsonDocument.Parse(await response.Content.ReadAsStringAsync(ct));
            var result = Aggregate(document.RootElement, market.ConditionId, start, now, windowMinutes);
            _log.LogInformation("Wallet-flow shadow: {Trades} trades, ${Volume:F2} volume, imbalance={Imbalance:+0.00;-0.00;0.00}",
                result.TradeCount, result.GrossVolumeUsd, result.FlowImbalance);
            return result;
        }
        catch (Exception ex)
        {
            _log.LogWarning("Wallet-flow shadow lookup failed: {Message}", ex.Message);
            return null;
        }
    }

    private WalletFlowReference Aggregate(JsonElement root, string conditionId, long start, long end, int windowMinutes)
    {
        var yesVolume = 0.0;
        var noVolume = 0.0;
        var largeVolume = 0.0;
        var tradeCount = 0;
        var largeCount = 0;
        var walletVolume = new Dictionary<string, double>(StringComparer.OrdinalIgnoreCase);
        var threshold = Math.Max(0, _config.WalletFlowLargeTradeUsd);

        if (root.ValueKind == JsonValueKind.Array)
        foreach (var trade in root.EnumerateArray())
        {
            if (!Text(trade, "conditionId").Equals(conditionId, StringComparison.OrdinalIgnoreCase)) continue;
            var timestamp = Number(trade, "timestamp");
            var size = Number(trade, "size");
            var price = Number(trade, "price");
            if (timestamp < start || timestamp > end || size <= 0 || price <= 0) continue;
            var notional = size * price;
            if (!double.IsFinite(notional) || notional <= 0) continue;

            var side = Text(trade, "side").ToUpperInvariant();
            var outcome = Text(trade, "outcome").ToUpperInvariant();
            var yesDirection = outcome == "YES" && side == "BUY" || outcome == "NO" && side == "SELL";
            var noDirection = outcome == "NO" && side == "BUY" || outcome == "YES" && side == "SELL";
            if (!yesDirection && !noDirection) continue;

            tradeCount++;
            if (yesDirection) yesVolume += notional;
            else noVolume += notional;
            var wallet = Text(trade, "proxyWallet").Trim();
            if (wallet.Length > 0)
                walletVolume[wallet] = walletVolume.GetValueOrDefault(wallet) + notional;
            if (notional >= threshold)
            {
                largeCount++;
                largeVolume += notional;
            }
        }

        var gross = yesVolume + noVolume;
        return new WalletFlowReference(
            windowMinutes,
            tradeCount,
            walletVolume.Count,
            gross,
            yesVolume,
            noVolume,
            yesVolume - noVolume,
            gross > 0 ? (yesVolume - noVolume) / gross : 0,
            gross > 0 ? walletVolume.Values.DefaultIfEmpty().Max() / gross : 0,
            largeCount,
            gross > 0 ? largeVolume / gross : 0,
            end);
    }

    private static string Text(JsonElement element, string property)
        => element.TryGetProperty(property, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString() ?? ""
            : "";

    private static double Number(JsonElement element, string property)
    {
        if (!element.TryGetProperty(property, out var value)) return 0;
        if (value.ValueKind == JsonValueKind.Number && value.TryGetDouble(out var number)) return number;
        return value.ValueKind == JsonValueKind.String &&
               double.TryParse(value.GetString(), System.Globalization.NumberStyles.Float,
                   System.Globalization.CultureInfo.InvariantCulture, out number)
            ? number
            : 0;
    }
}
